File: ui/backend/routes/environments.py
"""Environment routes — CRUD for named variable sets."""

from typing import Optional

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from models import Environment
from services.storage import (
    delete_environment,
    get_environment,
    list_environments,
    save_environment,
)

router = APIRouter()


class CreateEnvironmentBody(BaseModel):
    name: str
    variables: dict[str, str] = Field(default_factory=dict)


class UpdateEnvironmentBody(BaseModel):
    variables: dict[str, str]


@router.get("")
def list_envs():
    """List all environments."""
    return [e.model_dump() for e in list_environments()]


@router.get("/{name}")
def get_env(name: str):
    """Get a single environment."""
    env = get_environment(name)
    if env is None:
        raise HTTPException(404, f"Environment '{name}' not found")
    return env.model_dump()


@router.post("")
def create_env(body: CreateEnvironmentBody):
    """Create a new environment."""
    if get_environment(body.name) is not None:
        raise HTTPException(409, f"Environment '{body.name}' already exists")
    env = Environment(name=body.name, variables=body.variables)
    save_environment(env)
    return env.model_dump()


@router.put("/{name}")
def update_env(name: str, body: UpdateEnvironmentBody):
    """Update an existing environment's variables."""
    existing = get_environment(name)
    if existing is None:
        raise HTTPException(404, f"Environment '{name}' not found")
    updated = Environment(name=existing.name, variables=body.variables)
    save_environment(updated)
    return updated.model_dump()


@router.delete("/{name}")
def delete_env(name: str):
    """Delete an environment."""
    if not delete_environment(name):
        raise HTTPException(404, f"Environment '{name}' not found")
    return {"status": "deleted", "name": name}

File: ui/tauri/src/commands/environments.rs
//! Environment commands — mirrors `ui/backend/routes/environments.py`.

use std::collections::HashMap;

use tauri::State;

use crate::storage::{self, Environment, WorkspaceState};

#[tauri::command]
#[specta::specta]
pub fn list_environments(state: State<'_, WorkspaceState>) -> Result<Vec<Environment>, String> {
    storage::list_environments(&state)
}

#[tauri::command]
#[specta::specta]
pub fn get_environment(
    state: State<'_, WorkspaceState>,
    name: String,
) -> Result<Environment, String> {
    storage::get_environment(&state, &name)?
        .ok_or_else(|| format!("Environment '{}' not found", name))
}

#[tauri::command]
#[specta::specta]
pub fn create_environment(
    state: State<'_, WorkspaceState>,
    name: String,
    variables: HashMap<String, String>,
) -> Result<Environment, String> {
    if storage::get_environment(&state, &name)?.is_some() {
        return Err(format!("Environment '{}' already exists", name));
    }
    let env = Environment { name, variables };
    storage::save_environment(&state, &env)?;
    Ok(env)
}

#[tauri::command]
#[specta::specta]
pub fn update_environment(
    state: State<'_, WorkspaceState>,
    name: String,
    variables: HashMap<String, String>,
) -> Result<Environment, String> {
    let existing = storage::get_environment(&state, &name)?
        .ok_or_else(|| format!("Environment '{}' not found", name))?;
    let updated = Environment {
        name: existing.name,
        variables,
    };
    storage::save_environment(&state, &updated)?;
    Ok(updated)
}

#[tauri::command]
#[specta::specta]
pub fn delete_environment(state: State<'_, WorkspaceState>, name: String) -> Result<(), String> {
    if !storage::delete_environment(&state, &name)? {
        return Err(format!("Environment '{}' not found", name));
    }
    Ok(())
}
